### ISSUE TITLE
Add error handling for windows RPC APIs (that use SEH)

### ISSUE THREAD

**Issue by ChrisGuzak** [microsoft/wil#49]
A feature request from internal windows teams.

RETURN_IF_RPC_FAILED/THROW_IF_RPC_FAILED/FAIL_FAST_IF_RPC_FAILED macros that would use a lambda to create a function scope that enables the use of the __try/__catch/__finally language extensions needed to consume RPC based APIs.

#define RETURN_IF_RPC_FAILED(hr) \
do \
{ \
HRESULT __hrRet = [&]() -> HRESULT \
{ \
RpcTryExcept \
{ \
HRESULT __hrRet = wil::verify_hresult(hr); \
if (FAILED(__hrRet)) \
{ \
__RETURN_HR_FAIL(__hrRet, #hr); \
} \
} \
RpcExcept(I_RpcExceptionFilter(RpcExceptionCode())) \
{ \
__RETURN_WIN32_FAIL(RpcExceptionCode(), #hr); \
} \
RpcEndExcept \
return S_OK; \
}(); \
} \
while (0, 0)

**Comment by jonwis** [comment]
My github-fu is weak, but I started https://github.com/microsoft/wil/compare/user/jonwis/rpc with the method-wrapper variant.

I'm not happy with it yet. Feels like I should be able to use the new `wistd::invoke` to take both lambdas and methods-with-parameters:

```c++
// Lambda form:
wil::call_rpc_nothrow([&] { return MyRpcMethod(m_connection, a, b, c); });

// Parameters form:
wil::call_rpc_nothrow(MyRpcMethod, a, b,c);
```

I'm also struggling on the throwing 'with result' form. Seems like I shouldn't have to specify the GUID, and some traits-magic could figure it out:

```c++
// Current
GUID g = wil::call_rpc<GUID>(MethodThatReturnsGuid, a, b, c);

// Preferred
GUID g = wil::call_rpc(MethodThatReturnsGuid, a, b, c);
```

**Comment by kennykerr** [comment]
Some inspiration from C++/WinRT perhaps. Have a look at [capture](https://github.com/microsoft/xlang/blob/318fcf462e24e894c2d605a08a1c249a0044ec42/src/tool/cppwinrt/strings/base_com_ptr.h#L157) and then [create_instance](https://github.com/microsoft/xlang/blob/318fcf462e24e894c2d605a08a1c249a0044ec42/src/tool/cppwinrt/strings/base_activation.h#L400) as an example of how to avoid wrapping every call in a lambda.

**Comment by dunhor** [comment]
See [`std::invoke_result`](https://en.cppreference.com/w/cpp/types/result_of), which may or may not already be present in `wistd`. That's one way to deduce the return type without having to explicitly declare it.

**Comment by sylveon** [comment]
It's not at the moment, but it can easily be added using wistd::__invoke_of or can be substituted by wistd::result_of.

**Comment by jonwis** [comment]
RPC typically only supports POD types, so this is probably an ok restriction. Thanks for the tips on result_of.

**Comment by ChrisGuzak** [comment]
@Kenny, capture is cool. one use case
```
        auto compileOnlyTest = []()
        {
            winrt::com_ptr<IClassFactory> factory;

            winrt::com_ptr<IUnknown> obj;
            winrt::check_hresult(factory->CreateInstance(nullptr, __uuidof(obj), obj.put_void()));
            obj = nullptr; // going to reuse
            obj.capture(factory, &IClassFactory::CreateInstance, nullptr);
        };
```

**Comment by jonwis** [comment]
Closed by #51
